### PR TITLE
[#36] Add support for cross-sheet local addresses

### PR DESCRIPTION
**Pull request by Adanteh**
No idea if what I'm doing here makes sense at all, but all tests are still good, and it passes the new test case that was failing originally.

### CODE REVIEW

File: xlcalculator/ast_nodes.py [hunk 1]
```diff
@@ -47,10 +47,15 @@ def __init__(self, namespace=None, ref=None, seen=None):
         self.seen = seen if seen is not None else []
         self.namespace = namespace if namespace is not None else xl.FUNCTIONS
         self.ref = ref
+        self.refsheet = ref.split("!")[0]
+        self.sheet = self.refsheet
```
Comment (strichter) [inline xlcalculator/ast_nodes.py:51]: Combine into single, multi-assginment statement.

Comment (strichter) [reply xlcalculator/ast_nodes.py:51]: Add new attributes in class declaration above. The usage of `refsheet` and `sheet` is exactely the opposite of what I would have thought, so it needs some documentation (at the class level).

File: xlcalculator/ast_nodes.py [hunk 2]
```diff
@@ -148,7 +152,9 @@ def eval(self, context):
             context.ranges[addr].value = data = func_xltypes.Array(range_cells)
             return data
 
-        return context.eval_cell(addr)
+        value = context.eval_cell(addr)
+        context.reset()
```
Comment (strichter) [inline xlcalculator/ast_nodes.py:156]: It is unclear to me why you need `reset()`. Are we setting `context.sheet` anywhere else? To me it looks like a new attribute you are introducing.

File: xlcalculator/evaluator.py [hunk 3]
```diff
@@ -27,10 +27,10 @@ def eval_cell(self, addr):
             raise RuntimeError(
                 f'Cycle detected for {addr}:\n- ' + '\n- '.join(self.seen))
         self.seen.append(addr)
+        self.sheet = addr.split("!")[0]
 
         return self.evaluator.evaluate(addr, self)
 
-
```
Comment (strichter) [inline xlcalculator/evaluator.py:33]: Two empty lines after classes are required. Undo this change please.

File: xlcalculator/ast_nodes.py (hunk 2 above)
Comment (strichter) [reply xlcalculator/ast_nodes.py:156]: Oh you change it in evaluator.py.

File: xlcalculator/evaluator.py [hunk 4]
```diff
@@ -27,10 +27,10 @@ def eval_cell(self, addr):
             raise RuntimeError(
                 f'Cycle detected for {addr}:\n- ' + '\n- '.join(self.seen))
         self.seen.append(addr)
+        self.sheet = addr.split("!")[0]
```
Comment (strichter) [inline xlcalculator/evaluator.py:30]: There is an asymmetry here between where you set `sheet` and then reset it. This makes it very hard to see where you enter and leave a context. I would be nice to have this assignment and the reset() in the same function.

File: xlcalculator/ast_nodes.py (hunk 1 above)
Comment (Adanteh) [reply xlcalculator/ast_nodes.py:51]: From what I understood ref always pointed towards the original cell we're evaluating?

File: xlcalculator/evaluator.py (hunk 4 above)
Comment (Adanteh) [reply xlcalculator/evaluator.py:30]: Ideally proper contextmanager would be used in general. Unfortunately I don't have so much time to spend on it for now, it's more an idea of what would work than MR than should be pulled as is.

Comment (bradbase) [reply xlcalculator/evaluator.py:30]: @adanteh, I think a proper contextmanager is worthwhile. That said, not being able to cross-sheets is a fundamental problem. This proposal appears to resolve the cross-sheet problem for now. I have attempted to address the suggestions from @strichter and will update this PR with those changes shortly.

@adanteh, would this change adequately address your use case until we can raise a contextmanager?
